File: drift/src/main/java/drift/com/drift/managers/SocketManager.java
package drift.com.drift.managers;


import android.os.Handler;
import android.os.Looper;

import com.google.gson.Gson;
import com.google.gson.JsonElement;

import org.phoenixframework.PhxChannel;
import org.phoenixframework.PhxMessage;
import org.phoenixframework.PhxSocket;

import java.util.Map;

import drift.com.drift.api.APIManager;
import drift.com.drift.helpers.LoggerHelper;
import drift.com.drift.model.Message;
import drift.com.drift.model.SocketAuth;
import kotlin.Unit;
import kotlin.jvm.functions.Function0;
import kotlin.jvm.functions.Function1;
import kotlin.jvm.functions.Function2;
import okhttp3.OkHttpClient;
import okhttp3.Response;


/**
 * Created by eoin on 15/08/2017.
 */

public class SocketManager {

    private static String TAG = SocketManager.class.getSimpleName();

    private static SocketManager _socketManager = new SocketManager();

    private final Gson gson = APIManager.generateGson();

    private PhxSocket socket;
    private PhxChannel channel;

    public static SocketManager getInstance() {
        return _socketManager;
    }

    public void disconnect() {
        try {

            if (socket != null && socket.isConnected()) {
                socket.disconnect();
            }

        } catch (Exception e) {
            LoggerHelper.logMessage(TAG, "Failed to disconnect");
        }
    }

    public boolean isConnected(){

        if (socket != null) {
            return socket.isConnected();
        }
        return false;
    }

    public void connect(final SocketAuth auth) {

        try {

            if (socket != null && socket.isConnected()) {
                socket.disconnect();
            }
            String url = getSocketURL(auth.orgId, auth.sessionToken);
            socket = new PhxSocket(url, null, new OkHttpClient());

            socket.setLogger(new Function1<String, Unit>() {
                @Override
                public Unit invoke(String s) {
                    LoggerHelper.logMessage("Drift Socket", s);
                    return Unit.INSTANCE;
                }
            });

            socket.onOpen(new Function0<Unit>() {
                @Override
                public Unit invoke() {

                    LoggerHelper.logMessage(TAG, "Connected");
                    channel = socket.channel("user:" + auth.userId, null);



                    channel.on("change", new Function1<PhxMessage, Unit>() {
                        @Override
                        public Unit invoke(PhxMessage phxMessage) {

                            LoggerHelper.logMessage(TAG, phxMessage.getEvent());

                            LoggerHelper.logMessage(TAG, phxMessage.getPayload().toString());

                            if (phxMessage.getPayload().containsKey("body")){

                                Object body = phxMessage.getPayload().get("body");
                                LoggerHelper.logMessage(TAG, "Body " + body.toString());

                                if (body instanceof Map<?, ?>){

                                    if (((Map) body).containsKey("object")){

                                        Object object = ((Map) body).get("object");
                                        LoggerHelper.logMessage(TAG, "Object " + object.toString());
                                        if (object instanceof Map<?, ?>) {

                                            if (((Map) object).containsKey("type") && ((Map) body).containsKey("data")) {
                                                Object type = ((Map) object).get("type");
                                                Object data = ((Map) body).get("data");


                                                LoggerHelper.logMessage(TAG, "Type " + type.toString());
                                                LoggerHelper.logMessage(TAG, "Data " + data.toString());
                                                if (type instanceof String && data instanceof Map<?,?>){

                                                    processEnvelopeData((String) type, (Map) data);
                                                    return Unit.INSTANCE;
                                                }

                                            }
                                        }
                                    }
                                }
                            }

                            LoggerHelper.logMessage(TAG, "Failed to parse envelope! " + phxMessage.getPayload());

                            return Unit.INSTANCE;
                        }
                    });


                    channel.join(null, null)
                    .receive("ok", new Function1<PhxMessage, Unit>() {
                        @Override
                        public Unit invoke(PhxMessage phxMessage) {
                            LoggerHelper.logMessage(TAG, "You have joined '" + phxMessage.getEvent() + "'");

                            return Unit.INSTANCE;
                        }
                    });

                    return Unit.INSTANCE;
                }
            });

            socket.onClose(new Function0<Unit>() {
                @Override
                public Unit invoke() {
                    LoggerHelper.logMessage(TAG, "Closed Socket");
                    return Unit.INSTANCE;
                }
            });


            socket.onError(new Function2<Throwable, Response, Unit>() {
                @Override
                public Unit invoke(Throwable throwable, Response response) {
                    throwable.printStackTrace();
                    LoggerHelper.logMessage(TAG, "Socket Error: " + response);


                    return Unit.INSTANCE;
                }
            });

            socket.connect();


        } catch (Exception e) {
            LoggerHelper.logMessage(TAG, "Failed to connect");
        }
    }

    private void processEnvelopeData(final String type, final Map<String, Object> data) {


        Handler mainHandler = new Handler(Looper.getMainLooper());

        Runnable myRunnable = new Runnable() {
            @Override
            public void run() {

                switch (type) {
                    case "MESSAGE":
                        JsonElement jsonData = gson.toJsonTree(data);

                        Message message = gson.fromJson(jsonData, Message.class);

                        if (message != null && message.contentType.equals("CHAT")) {
                            LoggerHelper.logMessage(TAG, "Received new Message");
                            PresentationManager.getInstance().didReceiveNewMessage(message);
                        }

                        break;
                    default:
                        LoggerHelper.logMessage(TAG, "Undealt with socket event type: " + type);
                        break;
                }

            }
        };
        mainHandler.post(myRunnable);
    }

    private String getSocketURL(int orgId, String socketAuth) {
        return "wss://" + String.valueOf(orgId) + "-" + String.valueOf(computeShardId(orgId)) + ".chat.api.drift.com/ws/websocket?session_token=" + socketAuth;
    }

    private static int computeShardId(int orgId){
        return orgId % 50; //WS_NUM_SHARDS
    }

}

File: drift/src/main/java/drift/com/drift/managers/SocketManager.kt
package drift.com.drift.managers


import android.os.Handler
import android.os.Looper
import org.phoenixframework.PhxChannel
import org.phoenixframework.PhxSocket
import drift.com.drift.api.APIManager
import drift.com.drift.helpers.LoggerHelper
import drift.com.drift.model.Message
import drift.com.drift.model.SocketAuth
import okhttp3.OkHttpClient


/**
 * Created by eoin on 15/08/2017.
 */

internal object SocketManager {

    private val TAG = SocketManager::class.java.simpleName

    private val gson = APIManager.generateGson()

    private var socket: PhxSocket? = null
    private var channel: PhxChannel? = null

    val isConnected: Boolean
        get() = socket?.isConnected ?: false

    fun disconnect() {
        try {

            if (socket?.isConnected == true) {
                socket?.disconnect()
            }

        } catch (e: Exception) {
            LoggerHelper.logMessage(TAG, "Failed to disconnect")
        }

    }

    fun connect(auth: SocketAuth) {

        try {

            if (socket?.isConnected == true) {
                socket?.disconnect()
            }
            val url = getSocketURL(auth.orgId, auth.sessionToken)
            socket = PhxSocket(url, null, OkHttpClient())

            socket?.logger = {
                LoggerHelper.logMessage("Drift Socket", it)
            }

            socket?.onOpen {
                LoggerHelper.logMessage(TAG, "Connected")
                channel = socket?.channel("user:" + auth.userId!!, null)



                channel?.on("change") { phxMessage ->
                    LoggerHelper.logMessage(TAG, phxMessage.event)

                    LoggerHelper.logMessage(TAG, phxMessage.payload.toString())

                    if (phxMessage.payload.containsKey("body")) {

                        val body = phxMessage.payload["body"]
                        LoggerHelper.logMessage(TAG, "Body " + body!!.toString())

                        if (body is Map<*, *>) {

                            if (body.containsKey("object")) {

                                val `object` = body["object"]
                                LoggerHelper.logMessage(TAG, "Object " + `object`!!.toString())
                                if (`object` is Map<*, *>) {

                                    if (`object`.containsKey("type") && body.containsKey("data")) {
                                        val type = `object`["type"]
                                        val data = body["data"]


                                        LoggerHelper.logMessage(TAG, "Type " + type!!.toString())
                                        LoggerHelper.logMessage(TAG, "Data " + data!!.toString())
                                        if (type is String && data is Map<*, *>) {
                                            processEnvelopeData(type, data as Map<String, Any>)
                                            return@on Unit
                                        }

                                    }
                                }
                            }
                        }
                    }
                    LoggerHelper.logMessage(TAG, "Failed to parse envelope! " + phxMessage.payload)
                }


                channel?.join(null, null)
                        ?.receive("ok") { phxMessage ->
                            LoggerHelper.logMessage(TAG, "You have joined '" + phxMessage.event + "'")
                        }
            }

            socket?.onClose {
                LoggerHelper.logMessage(TAG, "Closed Socket")
            }

            socket?.onError { throwable, response ->
                throwable?.printStackTrace()
                LoggerHelper.logMessage(TAG, "Socket Error: $response")
            }

            socket?.connect()


        } catch (e: Exception) {
            LoggerHelper.logMessage(TAG, "Failed to connect")
        }

    }

    private fun processEnvelopeData(type: String, data: Map<String, Any>) {


        val mainHandler = Handler(Looper.getMainLooper())

        val myRunnable = Runnable {
            when (type) {
                "MESSAGE" -> {
                    val jsonData = gson.toJsonTree(data)

                    val message = gson.fromJson(jsonData, Message::class.java)

                    if (message != null && message.contentType == "CHAT") {
                        LoggerHelper.logMessage(TAG, "Received new Message")
                        PresentationManager.didReceiveNewMessage(message)
                    }
                }
                else -> LoggerHelper.logMessage(TAG, "Undealt with socket event type: $type")
            }
        }
        mainHandler.post(myRunnable)
    }

    private fun getSocketURL(orgId: Int, socketAuth: String?): String {
        return "wss://" + orgId.toString() + "-" + computeShardId(orgId).toString() + ".chat.api.drift.com/ws/websocket?session_token=" + socketAuth
    }

    private fun computeShardId(orgId: Int): Int {
        return orgId % 50 //WS_NUM_SHARDS
    }
}
